reverse_proxy_handler.cc

```diff
@@ -51,88 +52,61 @@ RequestHandler::Status ReverseProxyHandler::Init(const std::string& uri_prefix,
-        return RequestHandler::Error;
-    }
-
-    std::cerr << "Got past connecting to remote_host!" << std::endl;
-
-    // Construct new request to send to remote host
-    // Example: Modify request from: /reverse_proxy/static/file1.txt
-    // to:                           /static/file1.txt
-    std::string remote_uri = request.uri();
-    remote_uri.erase(0, original_uri_prefix.size());
-
-    if (remote_uri.empty()) {
-      remote_uri = "/";
-    }
-
-    std::string remote_request
-      = "GET " + remote_uri + " HTTP/1.1" + "\r\n\r\n";
-
-    std::cerr << "Sending remote_request: " << remote_request << std::endl;
-
-    boost::asio::write(socket, boost::asio::buffer(remote_request));
-
-    // Based on: https://github.com/UCLA-CS130/Mr.-Robot-et-al./blob/c9b064c68cd4bc1ae6b5c012db59eae9cb8b946d/lightning_server.cc#L26
-    const int MAX_BUFFER_LENGTH = 8192;
-    char response_buffer[MAX_BUFFER_LENGTH];
-    memset(response_buffer, 0, MAX_BUFFER_LENGTH);
+// Helper for 302-redirects
+// Returns the value of the Location header,
+// e.g., "www.exmple.com" for "Location: www.example.com"
+std::string ReverseProxyHandler::getLocationHeaderValue(const std::string& response_buffer_string) {
+    // Grab the Location header line
+    int start_location = response_buffer_string.find("Location:");
+    int end_location = response_buffer_string.find_first_of("\r\n", start_location);
+    std::string location_line = response_buffer_string.substr(start_location, end_location);
 
-    socket.read_some(boost::asio::buffer(response_buffer), ec);
+    boost::tokenizer<boost::char_separator<char>> tokens
+        = tokenGenerator(location_line, " ");
 
-    switch (ec.value()) {
-      case boost::system::errc::success:
-        std::cout << "remote_host's response: \n" << response_buffer << std::endl;
-        break;
-      default:
-        std::cout << "Error reading from remote_host, error code: " << ec << std::endl;
-        return RequestHandler::Error;
+    // Split the location line to get the value
+    std::string location_value;
+    int i = 0;
+    for (auto cur_token = tokens.begin();
+         cur_token != tokens.end();
+         cur_token++,
+         i++) {
+        if (i == 1) {
+            location_value = *cur_token;
+            break;
+        }
     }
 
-    // Handle response code from remote_host.
-    // remote_host's ResponseCode => our ResponseCode cases:
-    //
-    // 200 => 200
-    // 302 => fetch-loop to 200 or 404 (upon not-found or max-retries)
-    // 404 => 404
-    // 500 => 404
+    return location_value;
+}
 
+// Helper for parsing response code from origin responses
+std::string ReverseProxyHandler::getRemoteResponseCode(const std::string& response_buffer_string) {
     // Response has the form:
     // HTTP/1.0 200 OK
     // Content-Length: 3497
     // Content-Type: text/plain
+    //
     // <body contents>
-
+    //
```

Comment:
This is actually a separator line, which is non-obvious from the diff: 

![image](https://cloud.githubusercontent.com/assets/4713937/23577103/9fac1900-006b-11e7-98f7-b11b6e451f59.png)